### PR TITLE
Feature/http list provider

### LINKED ISSUE #47: Multiple RPCs to send transaction in parallel
Since one RPC node could be in not-synced state and that is why will not propagate transaction sent by the relayer it makes sense to provide an ability to configure several RPC endpoints and send transactions to them in parallel.

### LINKED ISSUE #50: Handle common JSON RPC error codes
See example here: https://github.com/omni/tokenbridge/blob/ff9f3fb7d6b4b8b484b4ef92427ce03411265714/oracle/src/services/HttpListProvider.js#L11

### LINKED ISSUE #56: "Internal error" with stack trace does not provide clue on the error scenario
With the relayer `1.2.1` the following message appeared in the logs:

```
Sep 14 16:17:12 relayer-mpv zk-zkbob-relayer-1/3b2ae199da3b  #033[32minfo#033[39m: Adding nullifier 11470810142313174456432862331998446141193071694015032832014961839710479633968 to OS {"timestamp":"2022-09-14T13:17:12.699Z"}
Sep 14 16:17:35 relayer-mpv zk-zkbob-relayer-1/3b2ae199da3b Error: Returned error: Internal error
Sep 14 16:17:35 relayer-mpv zk-zkbob-relayer-1/3b2ae199da3b     at Object.ErrorResponse (/app/node_modules/web3-core-helpers/lib/errors.js:28:19)
Sep 14 16:17:35 relayer-mpv zk-zkbob-relayer-1/3b2ae199da3b     at /app/node_modules/web3-core-requestmanager/lib/index.js:300:36
Sep 14 16:17:35 relayer-mpv zk-zkbob-relayer-1/3b2ae199da3b     at XMLHttpRequest.request.onreadystatechange (/app/node_modules/web3-providers-http/lib/index.js:98:13)
Sep 14 16:17:35 relayer-mpv zk-zkbob-relayer-1/3b2ae199da3b     at XMLHttpRequestEventTarget.dispatchEvent (/app/node_modules/xhr2-cookies/dist/xml-http-request-event-target.js:34:22)
Sep 14 16:17:35 relayer-mpv zk-zkbob-relayer-1/3b2ae199da3b     at XMLHttpRequest._setReadyState (/app/node_modules/xhr2-cookies/dist/xml-http-request.js:208:14)
Sep 14 16:17:35 relayer-mpv zk-zkbob-relayer-1/3b2ae199da3b     at XMLHttpRequest._onHttpResponseEnd (/app/node_modules/xhr2-cookies/dist/xml-http-request.js:318:14)
Sep 14 16:17:35 relayer-mpv zk-zkbob-relayer-1/3b2ae199da3b     at IncomingMessage.<anonymous> (/app/node_modules/xhr2-cookies/dist/xml-http-request.js:289:61)
Sep 14 16:17:35 relayer-mpv zk-zkbob-relayer-1/3b2ae199da3b     at IncomingMessage.emit (node:events:525:35)
Sep 14 16:17:35 relayer-mpv zk-zkbob-relayer-1/3b2ae199da3b     at endReadableNT (node:internal/streams/readable:1358:12)
Sep 14 16:17:35 relayer-mpv zk-zkbob-relayer-1/3b2ae199da3b     at processTicksAndRejections (node:internal/process/task_queues:83:21)
Sep 14 16:17:42 relayer-mpv zk-zkbob-relayer-1/3b2ae199da3b  #033[32minfo#033[39m: SENT WORKER: Job 9: processing... {"timestamp":"2022-09-14T13:17:42.714Z"}
Sep 14 16:17:43 relayer-mpv zk-zkbob-relayer-1/3b2ae199da3b  #033[34mdebug#033[39m: SENT WORKER: Job 9: Transaction 0xcad57e633453e26046d4c3edddfec45474777b9ff1ae1277892a2ec01353114b was successfully mined at block 33096066 {"timestamp":"2022-09-14T13:17:42.931Z"}
```

Although the stack trace is provided, there is no any clue which sequence of actions lead to the error. Consider to improve logging to help investigation of such issues in the future.

### LINKED ISSUE #70: Handle incorrect RPC response on getTransactionReceipt
When sent worker tries to fetch a tx by hash, in some cases it gets an error:
```
Error: Returned error: Internal error
at Object.ErrorResponse (/app/node_modules/web3-core-helpers/lib/errors.js:28:19)
at /app/node_modules/web3-core-requestmanager/lib/index.js:300:36
at XMLHttpRequest.request.onreadystatechange (/app/node_modules/web3-providers-http/lib/index.js:98:13)
at XMLHttpRequestEventTarget.dispatchEvent (/app/node_modules/xhr2-cookies/dist/xml-http-request-event-target.js:34:22)
at XMLHttpRequest._setReadyState (/app/node_modules/xhr2-cookies/dist/xml-http-request.js:208:14)
at XMLHttpRequest._onHttpResponseEnd (/app/node_modules/xhr2-cookies/dist/xml-http-request.js:318:14)
at IncomingMessage.<anonymous> (/app/node_modules/xhr2-cookies/dist/xml-http-request.js:289:61)
at IncomingMessage.emit (node:events:525:35)
at endReadableNT (node:internal/streams/readable:1358:12)
```
It leads to broken internal state and freezes txs in UI, while they are mined.
Relayer has to to restart such jobs when they fail.

### PR DESCRIPTION
**Pull request by LyzeOfKiel**
Changes:
* Added `HttpListProvider` and `RedundantHttpListProvider` analogous to `tokenbridge` project. `RPC_URL` env now can contain several RPC URLs. `RELAYER_TX_REDUNDANCY` env flag added to explicitly use `RedundantHttpListProvider` if needed.
Closes #47
Closes #50
* Minor simplifications in `checkMined` function used by `sentTxWorker`. Also, added a warn log when getting transaction receipt fails.
Closes #70
Closes #56 
* Switched to LTS `node:18` in Dockerfile.
* Added `yarn install --production` in Dockerfile to prune `devDependencies` after TypeScript build is done. It reduces the final image size by ~100 MB.

### CODE REVIEW

File: zp-relayer/workers/sentTxWorker.ts [hunk 1]
```diff
@@ -64,16 +64,17 @@ export async function createSentTxWorker<T extends EstimationType>(gasPrice: Gas
     for (let i = prevAttempts.length - 1; i >= 0; i--) {
       const txHash = prevAttempts[i][0]
       logger.info('Verifying %s ...', txHash)
-      tx = await web3.eth.getTransactionReceipt(txHash)
-      if (tx) break
+      try {
+        tx = await web3.eth.getTransactionReceipt(txHash)
+      } catch (e) {
+        logger.warn('Cannot get tx receipt for %s; Error: %s', txHash, (e as Error).message)
```
Comment (akolotov) [inline zp-relayer/workers/sentTxWorker.ts:70]: should the `warn` log message really contain "Error" in the text?

File: zp-relayer/workers/sentTxWorker.ts [hunk 2]
```diff
@@ -64,16 +64,17 @@ export async function createSentTxWorker<T extends EstimationType>(gasPrice: Gas
...
+        throw e
```
Comment (akolotov) [inline zp-relayer/workers/sentTxWorker.ts:71]:
```suggestion
         // Exception should be caught by `withLoop` to re-run job
        throw e
```

File: zp-relayer/workers/sentTxWorker.ts (hunk 1 above)
Comment (LyzeOfKiel) [reply zp-relayer/workers/sentTxWorker.ts]: Changed to `Error` to `RPC response`

File: zp-relayer/workers/sentTxWorker.ts (hunk 2 above)
Comment (LyzeOfKiel) [reply zp-relayer/workers/sentTxWorker.ts:72]: Updated comment

File: zp-relayer/utils/constants.ts [hunk 3]
```diff
@@ -11,6 +11,12 @@ const constants = {
   POOL_ID_SIZE: 6,
   ZERO_ADDRESS: '0x0000000000000000000000000000000000000000',
   INIT_ROOT: '11469701942666298368112882412133877458305516134926649826543144744382391691533',
+  RETRY_CONFIG: {
+    retries: 2,
+    factor: 1.4,
+    maxTimeout: 360000,
```
Comment (k1rill-fedoseev) [inline zp-relayer/utils/constants.ts:17]: 6 minutes seems a bit too much.
maybe 60 seconds?

**Review by k1rill-fedoseev — APPROVED** [review]

File: zp-relayer/services/providers/HttpListProvider.ts [hunk 4]
```diff
@@ -1,59 +1,45 @@
...
+    super(urls[0], options)
     this.currentIndex = 0
     this.lastTimeUsedPrimary = 0
+
+    this.urls = urls
+  }
+
+  private updateUrlIndex(index: number) {
+    this.currentIndex = index
+    this.host = this.urls[this.currentIndex]
```
[the remark this answers is not in the document]
Comment (LyzeOfKiel) [reply zp-relayer/services/providers/HttpListProvider.ts:35]: Tried to follow the interface of original `web3` provider. If it is not used internally by `web3` then we can remove it. Wdyt?

**Review by akolotov — APPROVED** [review]

File: zp-relayer/utils/constants.ts (hunk 3 above)
Comment (LyzeOfKiel) [reply zp-relayer/utils/constants.ts]: Updated to 60 seconds

File: zp-relayer/services/providers/HttpListProvider.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (k1rill-fedoseev) [reply zp-relayer/services/providers/HttpListProvider.ts:35]: Ok, not a problem, let's keep it

**Pull request merged**